### PR TITLE
Throttle and measure case imports

### PR DESCRIPTION
**Pull request by dannyroberts**
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/SAASP-10196

Review commit by commit. Changes included (more details in the commits):
- Count local form submissions (i.e. case importer, etc.) toward submission rate limits (but don't apply the limits to these requests)
- Throttle case importer fairly and only when servers under high load
- Have case importer send metrics that let us estimate typical import times for N cases

### CODE REVIEW

**Comment by dannyroberts** [comment]
Tested a simple upload locally and will put this on staging, but assuming that works as intended I'd like to deploy this today or tomorrow in time for an upcoming big import!

**Review by snopoke — APPROVED** [review]
What do you think about changing the implementation of this by extending the `_Importer` and adding the timing etc. to the subclass by wrapping parent class methods?

Just thinking that it would isolate the timing etc from the actual import code more.

**Comment by snopoke** [comment]
Also curious if you plan to rate limit any other internal submissions and if so whether it might make more sense to do all internal rate limiting inside `submit_form_locally` (though I can see some challenges with doing it there)

File: corehq/apps/case_importer/do_import.py [hunk 1]
```diff
@@ -119,22 +126,40 @@ def commit_caseblocks(self):
...
+            # and the amount of load that the requests from this import put on the databases.
+            # The amount of time to wait, during a high submission period
+            # and while this project is using up more than its fair share
+            # should be proportional to this heuristic.
+            # For a fully throttled domain, this will up to double
+            # the amount of time the case import takes
+            datadog_counter(
+                'commcare.case_importer.import_delayed_duration',
+                self._last_submission_duration, tags=[
+                    'domain:{}'.format(self.domain)
```
Comment (millerdev) [inline corehq/apps/case_importer/do_import.py]: Would it make sense to use a bucket counter here so we can see an approximate distribution of delay duration? I believe adding that would simply entail adding a `duration:<value_bucket>` tag.

Related: did you intend to pass `self._last_submission_duration` as the counter increment value? (It's not always an integer value, is it?)

**Review by millerdev — APPROVED** [review]

File: corehq/apps/case_importer/do_import.py (hunk 1 above)
Comment (dannyroberts) [reply corehq/apps/case_importer/do_import.py]: I didn't realize count only accepted integer values. I was intending to use this to make a bar graph where the height of the bar is the sum of the lengths of the delays during that period, so we I can validate that the throttling is achieving the goals of (1) being more aggressive when server load is high (2) staying completely out of the way when server usage is low.

But I can't find any other examples of us doing this with float values, so I'll assume you're right and this doesn't work the way I want it to. I'll change this to counting the number of delays and bucketing by duration like you suggested.

**Comment by dannyroberts** [comment]
@snopoke 

> What do you think about changing the implementation of this by extending the _Importer and adding the timing etc. to the subclass by wrapping parent class methods?
>
> Just thinking that it would isolate the timing etc from the actual import code more.

That's an interesting suggestion. I'll take a pass at what that might look like and report back.

> Also curious if you plan to rate limit any other internal submissions and if so whether it might make more sense to do all internal rate limiting inside submit_form_locally (though I can see some challenges with doing it there)

That's a good question. I think at the moment I'm only interested in case importer, so can have a clean mental model of the scope of the change while I iterate on it, but if this approach is a great success, I'll probably want to see how I can naturally extend it to other sources of local submissions.

**Comment by dannyroberts** [comment]
@snopoke I took your suggestion in my last commit. I _think_ it's cleaner? It's a tradeoff that it's a little more complex to reason about (true about class hierarchies in general), but it does nicely separate the variables and logic related to metrics and throttling from the core implementation.

**Comment by dannyroberts** [comment]
I tested the changes on staging and the changes were just the things suggested by reviewers, so I'm going to consider this approved and merge

**Comment by dannyroberts** [comment]
(Normally I'd try to get a final review / confirmation, but a lot of people are off now this is fairly time sensitive.)

**Comment by snopoke** [comment]
@dannyroberts thanks for the follow up.

**Pull request merged**